### PR TITLE
BS-273 Add backup option, fix README, add some docstrings

### PR DESCRIPTION
**Pull request by mpanarin**
Migration file now supports adding a backup operation

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -26,5 +26,5 @@ follows::
   $ virtualenv -p YOUR_PYTHON env
   $ source env/bin/activate
   $ pip install -e .
-  $ pip install pytest
+  $ pip install pytest mock
```
Comment (simahawk) [inline README.rst]: use test requirements and change here w/ `pip install -e .[test]`
https://stackoverflow.com/questions/29870629/pip-install-test-dependencies-for-tox-from-setup-py

File: marabunta/model.py [hunk 2]
```diff
@@ -27,14 +28,73 @@ def versions(self):
 
 class MigrationOption(object):
 
-    def __init__(self, install_command=None, install_args=None):
+    def __init__(self, install_command=None, install_args=None, backup=None):
+        """Class for an options block in a migration.
```
Comment (simahawk) [inline marabunta/model.py]: We know is "class". Just say "Option block in a migration" ;)

File: marabunta/model.py [hunk 3]
```diff
@@ -27,14 +28,73 @@ def versions(self):
...
+        """
         self.install_command = install_command or u'odoo'
         self.install_args = install_args or u''
+        self.backup = backup
+
+
+class MigrationBackupOption(object):
+
+    def __init__(self, command, command_args, ignore_if, stop_on_failure=True):
+        """Backup option in migration class.
```
Comment (simahawk) [inline marabunta/model.py]: same

File: marabunta/model.py [hunk 4]
```diff
@@ -27,14 +28,73 @@ def versions(self):
...
+
+        Migration allows using a backup command in order to perform specific
+        commands(unless explicitly opted-out) before the migration step.
```
Comment (simahawk) [inline marabunta/model.py]: space missing

File: marabunta/model.py [hunk 5]
```diff
@@ -27,14 +28,73 @@ def versions(self):
...
+
+        :param command: Backup command to execute
+        :type command: String
+        :param command_args: Arguments for a backup command
+        :type command_args: String
+        :param ignore_if: A command, that is evaluates
```
Comment (simahawk) [inline marabunta/model.py]: *evaluated

File: marabunta/model.py [hunk 6]
```diff
@@ -265,3 +348,57 @@ def execute(self, log):
 
     def __repr__(self):
         return u'Operation<{}>'.format(' '.join(self.command))
+
+
+class SilentOperation(Operation):
+    """Subclass of an Operation class for operations that do not require
```
Comment (simahawk) [inline marabunta/model.py]: `"""Operation that does not require logging or interactivity."""`

**Comment by mpanarin** [comment]
@simahawk AFAIU those docs are supposed to be generated automatically with sphinx D:

@oleksandrpaziuk if backup fails - the only reason is your command. So debug it )

File: tests/test_backup.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016-2017 Camptocamp SA
```
Comment (TDu) [inline tests/test_backup.py]: Dates do not seem right for a new file

File: marabunta/model.py [hunk 8]
```diff
@@ -44,35 +103,43 @@ def __init__(self, number, options):
...
     def is_noop(self):
+        """Check if version is a no operation version.
+        """
         has_operations = [mode.pre_operations or mode.post_operations
                           for mode in self._version_modes.values()]
         has_upgrade_addons = [mode.upgrade_addons or mode.remove_addons
                               for mode in self._version_modes.values()]
-        noop = (not has_operations and not has_upgrade_addons)
+        has_backup = self.backup
+        noop = not any((has_backup, has_upgrade_addons, has_operations))
```
Comment (guewen) [inline marabunta/model.py]: A noop is a version where we don't have any addons to install/upgrade and no operations → there is no point to run a backup for a noop version. We should not need to write 

```
- 10.1.0:
   backup: false
```

Only to disable the backup on a noop version. `is_noop` should not be changed then. However, when `is_noop` is True, the backup should not be performed.

Also, if we have 3 consecutive versions that needs a backup, we only need a backup before the first (take care of the case where this is a new database).

I see it this way:

In the runner (https://github.com/camptocamp/marabunta/blob/8b5192076079ce8d3ecacec043cc014902a75ce4/marabunta/runner.py#L79), before executing the versions, loop through all the versions and if at least one version needs a backup and the global backup option is configured, then run it.

File: marabunta/model.py [hunk 9]
```diff
@@ -44,35 +103,43 @@ def __init__(self, number, options):
         self.number = number
         self._version_modes = {}
         self.options = options
+        self.backup = False
```
[the remark this answers is not in the document]
Comment (mpanarin) [reply marabunta/model.py]: no, I set it further in `_parse_version`. This is only declaration

**Comment by mpanarin** [comment]
@simahawk @guewen @TDu please review the latest changes, as specification a little bit changed )

File: marabunta/runner.py [hunk 10]
```diff
@@ -28,8 +28,16 @@ def __init__(self, config, migration, database, table):
         # we don't want to do it again for another version
         self.upgraded_addons = set()
 
-    def log(self, message):
-        print_decorated(u'migration: {}'.format(message))
+    def log(self, message, decorated=True, stdout=True):
```
Comment (simahawk) [inline marabunta/runner.py]: I don't see any of these new options used. Am I wrong?

Comment (mpanarin) [reply marabunta/runner.py]: As run of the backup command was moved from version to migration - we need those options in execution of the command here:
https://github.com/camptocamp/marabunta/blob/d949b0974d20a2f3d9e5df02aaef56d2769ea70a/marabunta/model.py#L342

**Comment by mpanarin** [comment]
@grindtildeath yes. You specify the command or a script in the migration options, and it will be run during the migration. It is extra execution and you can use it for anything you want additionally to backing up, yes

**Comment by guewen** [comment]
> I just don't understand how is the backup really done as psycopg2 only appears in setup.py requirements.
> As I understand it though, it's more kind of an "extra execution" feature than strictly a backup option ?

Yep, it's agnostic regarding *how and with what tools* you run the backup. In our case, it should be only a `curl` command towards dump-bag, rather than an execution of pg_dump in the container.

**Pull request merged**